### PR TITLE
Canonicalize partial connects

### PR DESCRIPTION
**Pull request by darthscsi**
Canonicalize partial connects which are doing truncation to separate truncation and connect

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -616,3 +616,37 @@ OpFoldResult AsNonPassivePrimOp::fold(ArrayRef<Attribute> operands) {
...
+
+    LogicalResult matchAndRewrite(PartialConnectOp op,
+                                  PatternRewriter &rewriter) const override {
+      auto destType =
+          op.getOperand(0).getType().cast<FIRRTLType>().getPassiveType();
+      auto srcType = op.getOperand(1).getType().cast<FIRRTLType>();
+      if (destType == srcType)
+        return failure();
+
+      auto srcWidth = srcType.getBitWidthOrSentinel();
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:636]: Plz add a comment about what this is doing so we can remember at a glance without studying the algorithm :-)

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 2]
```diff
@@ -616,3 +616,37 @@ OpFoldResult AsNonPassivePrimOp::fold(ArrayRef<Attribute> operands) {
...
+      auto destWidth = destType.getBitWidthOrSentinel();
+
+      if (destType.isa<IntType>() && srcType.isa<IntType>() && srcWidth > 0 &&
+          destWidth > 0 && destWidth < srcWidth) {
+        auto shortened = rewriter.createOrFold<TailPrimOp>(
+            op->getLoc(), destType, op.getOperand(1), srcWidth - destWidth);
+        rewriter.createOrFold<ConnectOp>(op->getLoc(), op.getOperand(0),
+                                         shortened);
+        rewriter.eraseOp(op);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:645]: I think you can use `rewriter.replaceOpWithOp<ConnectOp>(op.getOperand(0), shortened);` which is terser and might be more efficient.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 3]
```diff
@@ -616,3 +616,37 @@ OpFoldResult AsNonPassivePrimOp::fold(ArrayRef<Attribute> operands) {
...
+      if (destType == srcType)
+        return failure();
+
+      auto srcWidth = srcType.getBitWidthOrSentinel();
+      auto destWidth = destType.getBitWidthOrSentinel();
+
+      if (destType.isa<IntType>() && srcType.isa<IntType>() && srcWidth > 0 &&
+          destWidth > 0 && destWidth < srcWidth) {
+        auto shortened = rewriter.createOrFold<TailPrimOp>(
+            op->getLoc(), destType, op.getOperand(1), srcWidth - destWidth);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:642]: Would it make sense to do the same thing for extends?  are they allowed?

**Review by lattner — APPROVED** [review]
Great idea.  This will make other canonicalizations happen more often - would it make sense to do the same thing with normal connects?

Note that lowering transformations should still work with these forms, as they can't depend on canonicalization running.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp (hunk 3 above)
Comment (darthscsi) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:642]: I don't believe they are allowed.

**Comment by darthscsi** [comment]
> Great idea. This will make other canonicalizations happen more often - would it make sense to do the same thing with normal connects?

I don't think we have normal connects of basic int types with different widths.  If we do, absolutely.

> Note that lowering transformations should still work with these forms, as they can't depend on canonicalization running.

Yea, the partial connect was causing a spurious wire to be generated.  It was correct, but ugly.

**Comment by seldridge** [comment]
> would it make sense to do the same thing with normal connects?

I think we should hold off on doing explicit truncation during canonicalization for _connect_ because this would be a deviation from the FIRRTL spec. That said, the SFC violates this and treats connect and partial connect the same here. The best fix here is to have Chisel start emitting blown out partial connects instead of blown out connects and then the SFC should be able to add stricter behavior.

What could happen for connect (and partial connect) is to put width extension into canonicalization as both allow connecting smaller to wider.

**Comment by lattner** [comment]
I think that connect allows implicit extension.  Implicit truncation is really dubious...

**Pull request merged**